File: publisher/src/main/java/com/isonah/publisher/service/StoryService.java
package com.isonah.publisher.service;

import com.isonah.publisher.model.entity.Story;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.kafka.core.KafkaTemplate;
import org.springframework.stereotype.Service;

import java.util.Optional;

import static com.isonah.publisher.Constants.KAFKA_TOPIC_NAME;

/**
 * @author ielksseyer
 */
@Service
public class StoryService {

    @Autowired
    private KafkaTemplate<String, Story> kafkaTemplate;

    private static final Logger LOGGER = LoggerFactory.getLogger(StoryService.class);

    public Story pushStory(Story story) {
        LOGGER.info("sending object {} to kafka", story);
        return Optional.ofNullable(story)
                       .map(this::sendMessage)
                       .orElse(null);

    }

    private Story sendMessage(Story story) {
        return Optional.ofNullable(story)
                       .map(s -> kafkaTemplate.send(KAFKA_TOPIC_NAME, story))
                       .map(s -> story)
                       .orElse(null);
    }
}

File: mystory/src/main/kotlin/com/isonah/mystory/service/StoryService.kt
package com.isonah.mystory.service

import com.isonah.mystory.model.entity.Story
import com.isonah.mystory.repository.StoryRepository
import org.slf4j.LoggerFactory
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.stereotype.Service
import reactor.core.publisher.Mono
import reactor.core.publisher.toFlux
import reactor.core.publisher.toMono


/**
 *
 *
 * @author ielksseyer
 */

@Service
class StoryService {

    val logger = LoggerFactory.getLogger(StoryService::class.java)

    @Autowired
    lateinit var storyRepository: StoryRepository

    fun create(story: Story): Mono<Story> {
        return story.toMono()
                .doOnNext { logger.info("creating entry for {}", it) }
                .map { storyRepository.save(it) }
                .doOnError { logger.error("could not save {} to database", it) }
                .doOnSuccess { logger.debug("entry {} successfly added", it)
                        .to}
    }

    fun findAll() = storyRepository.findAll().toFlux()
}